tests/README.md

```diff
@@ -0,0 +1,1510 @@
+skipped on the CI server which leads to the final question if test is ever executed.
+
+
+```php
+<?php
+	/**
+	 * Test member agreement
+	 * @group IL_Init
+	 */
+	public function testMemberAgreement()
+	{
+		include_once 'Services/Membership/classes/class.ilMemberAgreement.php';
+
+		global $ilDB;
+
+
+		$agree = new ilMemberAgreement(9999,8888);
+		$agree->read();
+		$agree->setAccepted(true);
+		$agree->save();
+
+		$agree = new ilMemberAgreement(9999,8888);
+		$agree->read();
+		$sta = $agree->isAccepted();
+		$this->assertEquals($sta,true);
+		$agree->delete();
+
+		$agree = new ilMemberAgreement(9999,8888);
+		$agree->read();
+		$sta = $agree->isAccepted();
+		$this->assertEquals($sta,false);
+
+		$sta = ilMemberAgreement::_hasAccepted(9999,8888);
+		$this->assertEquals($sta,false);
+
+		$agree = new ilMemberAgreement(9999,8888);
+		$agree->read();
+		$agree->setAccepted(true);
+		$agree->save();
+
+		$sta = ilMemberAgreement::_hasAgreementsByObjId(8888);
+		$this->assertEquals($sta,true);
+
+		$sta = ilMemberAgreement::_hasAgreements();
+		$this->assertEquals($sta,true);
+
+		ilMemberAgreement::_deleteByUser(9999);
+	}
+```
+
+<a name="testing-php-behaviour--wrong-class"></a>
+#### Testing PHP behaviour / Wrong class
+The Button test contains tests which belongs into another class and testing PHP behaviour.
+In the case of the *test_implements_factory_interface* test the only thing which is tested is the *Factory* class
+which is living in the "\ILIAS\UI\Implementation\Component\Button\" namespace. In addition the only
+tested logic is the instantiation of the concrete button implementations.
+
+Furthermore this test checks 4 different things:
+- Standard button implements Standard interface ?
+- Primary button implements Primary interface ?
+- Close button implements Close interface ?
+- Shy button implements Shy interface ?
+
+```php
+<?php
+class ButtonTest extends ILIAS_UI_TestBase {
+
+	public function getButtonFactory() {
+		return new \ILIAS\UI\Implementation\Component\Button\Factory();
+	}
+
+	public function test_implements_factory_interface() {
+		$f = $this->getButtonFactory();
+
+		$this->assertInstanceOf("ILIAS\\UI\\Component\\Button\\Factory", $f);
+		$this->assertInstanceOf
+			( "ILIAS\\UI\\Component\\Button\\Standard"
+			, $f->standard("label", "http://www.ilias.de")
+			);
+		$this->assertInstanceOf
+			( "ILIAS\\UI\\Component\\Button\\Primary"
+			, $f->primary("label", "http://www.ilias.de")
+			);
+		$this->assertInstanceOf
+			( "ILIAS\\UI\\Component\\Button\\Close"
+			, $f->close()
+			);
+		$this->assertInstanceOf
+			( "ILIAS\\UI\\Component\\Button\\Shy"
+			, $f->shy("label", "http://www.ilias.de")
+			);
+	}
+}
+```
+
+<a name="example-solution"></a>
+##### Example solution
+The proposed solution would be to remove the test entirely because they test if the new keyword works. However, if the author wishes to keep the
```

Comment:
I do not get that. Sure, we can talk about the naming of the test class (should probably be "ButtonFactoryTest", however, the test itself does make sense to me. If you are coding test driven, you start by implementing those tests you see here. If you then wire up your factories incorrectly (which happens extremely fast, especially if you have large factories, yes I am looking at you glyphs...) those tests here will fail.

Reply:
I totally agree to write the test before implementation, however I am not sure if the test really tests something. There is only one line of code or did I miss something ?

Reply:
This test is not structured according to tripple A, so you might just have missed the calls to the factory methods in the `assertInstanceOf`-calls. The test indeed tests that these return instances implementing the specified interfaces, or I missed something =)

It could well be split into multiple test methods according to tripple A.

Reply:
Oh yes, I will change it accordingly :)

Thats correct :). However there is no logic to test or have I missed something ?
If the method returns it will be 100% that button.
The test which are concerned about the button will be in the test classes of the button implementations and the handling of the exceptions of these buttons in the consumers of the Factory I guess.